### ISSUE TITLE
Openworld: Design and implement item-room association structs

### ISSUE THREAD

**Issue by Justin08784** [uchicago-cs/chiventure#1035]
Currently, `roomspec_t` has an `item_hash_t *items` field for storing items for that specific room type, but items spawn in same quantities for every instance of the room generated. We would like to allow game authors to –– per room –– customize item occurrence probability and possible quantity ranges for the item. This allows openworld to generate a much more diverse (and possibly exciting) map. The original `item_hash_t` functionality can be preserved with an occurrence probability of 1 (guaranteed spawn) and a quantity range of 1-1 (only 1 item). 

e.g. Iron ingots have a random chance of spawning in minecraft village chests. When they do, their quantity falls in the range of 1-3. 

We expect to create 1) a wiki-based design document and 2) a new branch (openworld/item-room-association) for these changes, which we hope to merge into `dev` using a PR.

**Comment by Justin08784** [comment]
Met with Mingyan on Zoom today. Discussed the following:

Plans:
- Introduce a new struct `itemspec_t` struct which specifies how an item spawns (probability, quantity ranges) for a specific `roomspec_t`.
- Add a `itemspec_t` hash table field to `roomspec_t`

Implementation problems:
How to generate multiple copies of each item?
Our initial idea was to generate a copy of the item in the hash, and add an identifying integer suffix in the item id (just like how every time you generate a room from roomspecA, its identifying number increments by 1 “roomA0”, “roomA1” ...). 

However, it turns out that [item-stacking](https://github.com/uchicago-cs/chiventure/issues/703 ) was already implemented. This solution enables items of identical item_id's to exist together.

**Comment by Justin08784** [comment]
Created design doc [wiki page](https://github.com/uchicago-cs/chiventure/wiki/Openworld:-Itemspec_t-Design-Doc).

**Comment by Justin08784** [comment]
copy_item_to_hash(), which creates deep copies of the given item and adds it to a given destination hash, now has item stacking enabled.

Added item stacking test in test_rooms.c.

**Comment by Justin08784** [comment]
Summarizing 6452354 and d07e5f3:
- Added levelspec_t struct and helpers
- Implemented load_items() function.
- Redefined MAX_RAND_ITEMS to 100

Tests for new functions are needed.
Note: 
- Please ignore the last three commits above this comment.
- Also, for some reason, wdl tests (which are unrelated to our branch?) are failing.

**Comment by Justin08784** [comment]
Using `git checkout .b 6ed552c41e88f959e6803644667bdd28b510d58c` in the root chiventure directory, I reset to the last working commit 6ed552c.

**Comment by Justin08784** [comment]
Cause of bug:
There is another load_items() function defined in WDL. The duplicate name causes the crash.

Fix:
Renamed OUR load_items() to generate_items().

**Comment by Justin08784** [comment]
Summary of 6b4d5ae:
- roomspec_to_room() now calls generate_items() instead of random_items()
- levelspec_new() explicitly sets its roomlevels to NULL; roomspec_new() explicitly sets its itemspecs to NULL (changed reflected in header file documentation)
- added a roomspec_to_room() test for when itemspecs have been initialized to 0 spawn_chance

Bugs in generate_items():
- `double spawn_chance` was constantly 0; [cause + fix](https://github.com/uchicago-cs/chiventure/commit/6b4d5ae8713eda0b8d6c029a0b0b11977d570b07#r51177283)
- `int spawn_num` was constantly 1; [cause + fix](https://github.com/uchicago-cs/chiventure/commit/6b4d5ae8713eda0b8d6c029a0b0b11977d570b07#r51177301)

**Comment by Justin08784** [comment]
Misc. Changes:
- The roomspec_init function is callocing space. It should NOT be doing this because it results in memory leaks
- Several struct_new functions contain 'if' internal checkpoints to make sure space for a certain field is allocated. If it is not, it prematurely returns NULL. However, some of them do not fully free before returning NULL, resulting in allocated space for the struct/other fields becoming lost (memory leaks).

**Comment by wmingyan** [comment]
Tests added:
- tests for the itemspec struct functions (init_itemspec(), itemspec_new(), itemspec_free()) have been added to test_structs.c
- tests for generate_items() have been added to test_autogenerate.c

**Comment by Justin08784** [comment]
Closing Statement:
We published the design document [wiki page](https://github.com/uchicago-cs/chiventure/wiki/Openworld:-Itemspec_t-Design-Doc) for itemspec-based item generation. The relevant pull request (#1106) has been approved and the new `itemspec_t` struct and the itemspec-based item generation function `generate_items()` (all along with relevant tests) has been merged into the dev branch. This marks the end of this issue.
